### ISSUE TITLE
sequential consistent

### ISSUE THREAD

**Issue by geronimo22** [cameron314/concurrentqueue#73]
Hello,

I was planning on using your queue but noticed that you mention that the queue doesn't maintain the order of elements -- i.e. FIFO?

I saw this under "not sequential consistent..."
https://github.com/cameron314/concurrentqueue#reasons-not-to-use

Can you please elaborate under which conditions is it not FIFO? or please me to some place where i can read about the ordering guarantees of your queue.

Thanks,

**Comment by cameron314** [comment]
Hi :-)

The queue is internally composed of a series of sub-queues, one for each producer (if explicit producer tokens are not used, the current thread is implicitly a producer). Elements put in by a given producer will all come out in the same order (FIFO). But there's no coordination between elements from other producers -- two elements put in by two different producers will come out in an undefined order relative to each other. Basically, you can think of the queue as a set of FIFO streams that get interspersed when dequeueing.

Aside from that, sequential consistency is a separate issue and again a stronger guarantee than what the queue offers -- even if elements across all producers were FIFO, sequential consistency is not necessarily implied. Sequential consistency means that all threads agree on a single ordering (interpretation of the sequence of events with respect to changes in state). Even if the elements on a given queue come out in perfect FIFO order, other memory effects in the program may not be observable in the same order with respect to the element order depending which thread is doing the observing at the time.

The README documents this, but perhaps not as clearly.

This was a design decision for improved performance, by the way. Linearizability come at a price ;-)

**Comment by geronimo22** [comment]
Thanks for your reply. I'm not I still follow the concept of sequential consistency.  I guess I will google the concept.

Do you know how your free differs from Boost lockfree queue in terms of sequential consistency?

**Comment by cameron314** [comment]
Sequential consistency (and the threaded memory model introduced in C++11) is generally not something you have to worry about unless you're writing low-level atomic stuff (e.g. using `std::atomic` directly). All sane higher-level constructs (data structures, synchronization primitives, etc.) tend to provide at least acquire-release semantics (which is all you need most of the time), including my queue. This ensures that all the effects of work done by a thread before it enqueues an element will be visible on another thread after it dequeues that element.

What causes more problems is linearizability -- most people expect a queue to have a defined order between all its elements, which is an assumption that mine breaks (the order is only defined between elements from a given producer).

Does this help?

**Comment by cameron314** [comment]
The memory ordering guarantees of Boost's queue don't seem to be documented, but looking at the source reveals that it provides acquire-release semantics, like my queue. This is quite common for lock-free data structures and is sufficient for almost all purposes.

**Comment by geronimo22** [comment]
The linearizability setup on your queue works for me. I still don't follow the concept of Sequential consistency or acquire release semantics -- just because I havent heard about them before. I will have to look up a good resource on it. Thanks for trying to explain though!

**Comment by caustik** [comment]
For some platforms (for example, ARM), you could do this in one thread:

varA = 1
varB = 2

and this in another thread:

if (varB == 2)
  if (varA != 1)
    printf("uhoh");

and witness the printf. For those platforms you need to use certain special
instructions to impose sequential consistency. x86 imposes this sort of
global ordering automatically. Things like a critical section handle this
for you automatically but it becomes an issue in lockfree algorithms
because they are designed to avoid memory synchronization instructions as
much as possible.

**Comment by cameron314** [comment]
Right. Even on x86 you don't get this automatically, since the processor is not the only source of memory effect re-orderings, but the compiler too (unless told otherwise, it can optimize happily away under the assumption that only a single thread is going to use the variables).

**Comment by poor1017** [comment]
Hi @cameron314 

Can I assume that `cameron314/readerwriterqueue` is definitely FIFO?

**Comment by poor1017** [comment]
@cameron314 

Thank you for your reply.

> the queue still doesn't offer sequentially consistent memory ordering with respect to other atomic operations

I do not understand this statement. 

In thread A:
```cpp
moodycamel::ConcurrentQueue<int> q;
q.enqueue(1);
q.enqueue(2);
```

In thread B:
```cpp
int item;
bool found = false;
found = q.try_dequeue(item);
std::cout << item << std::endl;
found = q.try_dequeue(item);
std::cout << item << std::endl;
```

Maybe I get:
```
2
1
```

Do you mean this? If not, can you teach me about `the queue still doesn't offer sequentially consistent memory ordering ` with some code?

**Comment by cameron314** [comment]
No, you will always get 1 then 2 out. But consider this:

Thread A:

    std::atomic<int> x;
    x = 17;
    //...
    x = 23;

Thread B:

    moodycamel::ConcurrentQueue<int> q;
    q.enqueue(1);
    if (x == 23) std::cout << "23" << std::endl;
    q.enqueue(2);

Thread C:

    int item;
    bool found = false;
    found = q.try_dequeue(item);
    std::cout << item << std::endl;
    found = q.try_dequeue(item);
    std::cout << item << std::endl;
    std::cout << x << std::endl;

It's possible for C to print 1 2 17 even if thread B prints 23. However, if the write to `x` was moved onto thread B for queueing, this would not be possible.

**Comment by poor1017** [comment]
@cameron314 

Thank you for your patience.

I think for thread C to print:
```
1
2
17
```
It's because in thread A, code is reordered.

I think the non sequentially consistent of the queue will cause the code in thread B reordered, because `if (x == 23) std::cout << "23" << std::endl;` is neither aquire nor releaase.

I am not sure if I understand your idea.

**Comment by cameron314** [comment]
Even if the code is not reordered (all the way down to the assembly instruction level), it's possible for a memory effect like this not to be seen by the different hardware cores at the same time, even if other (unrelated) memory effects are propagated.

Here the reads and writes to `x` are implicitly sequentially consistent, but the reads and writes done by the queue are not, so there's no global memory ordering.

However, enqueueing is a release operation for writes done on thread B, and dequeueing is an acquire operation which guarantees thread A will see the preceding writes done by thread B when it gets an element out. But this isn't the same as a total ordering across all threads for all operations.
